### DISCUSSION TITLE
General principles

### DISCUSSION THREAD

**Discussion by maylad31** [boris-kz/CogAlg#59] [General]
Here are my thoughts,
1) From what I understand, at each level, you cluster the input into patterns and these patterns serve as input for the next level. 

2) I agree that neural networks take a lot of iterations to generate accurate representations and in neural networks With each added layer,  the output that ultimately drives learning contains exponentially smaller fraction of original information. Your cross comparison is not random and and easy to refine as well.

3) However I still would like to fully understand the process of cross-comparison and feedback. It has been a bit difficult for me to understand as mostly I have worked with deep learning and less with conventional techniques. As per my understanding, you add unique set of operations at each level, compare the inputs and cluster them into patterns such that derivatives match. You have two terms loss and match that are clear by names  (I mean you would like to minimise the loss) but I am not clear about the exact process.

**Comment by boris-kz** [comment]
Thanks Mayank! 

- Cross-comp is the first stage of every level, starting with adjacent pixels. Please see cross_comp in line_patterns or comp_pixel in frame_blobs.
- I am selecting for match, not minimizing miss (loss). In this scheme, miss is a higher-derivation representation of the input, the value of that representation is borrowed from co-derived match (contrast value), please see   form_adjacent_M_(Pm_) in line_patterns,
- re feedback, etc. my reply to related questions may help: 

> Based on my understanding, the core of your algorithm finds patterns by
calculating the parameters: loss d and match m . You have defined m as the
inverse deviation of loss, thus with more loss, the less the match and vice
versa. 

Right, this is an indirect definition of match, used on the 1st stage of vision because correlation between intensity of reflected light and stability of reflecting object is very low.
More accurate direct definition of match is compression of represented magnitude: mag two inputs - mag their derivatives. Please see part 1 of my intro: http://www.cognitivealgorithm.info/2015/01/cognitive-algorithm-hierarchical-search.html
 
> Though I do not understand the use of bilateral match and the
mathematical intuition behind it. 

Think of it as a 1D version of 2D 3x3 kernel: match is defined for central pixel relative to surrounding pixels.

> Further, I could not understand the feedback mechanism. I only briefly
skimmed through the code and believe this has not been integrated yet, while
constants are passed as parameters for now, which later I believe will be
passed on to the lower layers from their successive layers.

Right.

> I fail to understand the mechanics of feedback system and some clarity here would be great.

Feedback is updating filters by higher-level average of input deviation from current value of these filters.
For ave m in 1D alg, it will be, roughly:

ave_deviation = (summed P.M / summed P.L)  / decay with distance from lower-level input
// computed on 2nd level: line_PPs

if abs ave_deviation > aave  # higher-order filter or filter-filter
lower-level ave_m +=  ave_deviation,
terminate Fspan   # same-filter span of inputs

I elaborate on feedback in part 3 of my intro, though it's a hard read

> Regarding feedback, I can see you have plans to use averages, like ave , from
successive layers to be used. But will these parameters be tuned per input
image?

For the same inputs, there will be optional reprocessing of their buffers in higher-level Ps: each P contains dert_, which is a buffer of their elements. This is related to imagination discussion in part 3, nothing in the code yet.

> I understand the algorithm gives a set of patterns and deep patterns(patterns
of patterns) for every input provided which aligns with your vision of instance
based learning. 
> Here I fail to understand how the learning works incase of a
new data point? Will the output of the algorithm will eventually be fed to some
sort of a clustering algorithm, which clusters them on the basis of patterns
found?

It's pipeline, each level is supposed to continuously receive new inputs from lower level, starting from raw images. 

> I would also like to know how you envision this to work with text? 

Text is labels of generalized representations: patterns. It should not be used as initial input because current hierarchy is empty: there are no patterns to associate these labels with. When we do have sufficiently general patterns, they can be associated with their labels, establishing corresponding short-cuts to the level that recognizes labels themselves. That's pretty distant from current work. 
Please see my part 0.  

> Will you be using a similar version of encoding on top of the algorithm to decode the text? 

I don't think so, it should be the same algorithm, just tuned to pay more attention (lower filters).
